data-prepper-plugins/opensearch-source/src/main/java/org/opensearch/dataprepper/plugins/source/opensearch/worker/NoSearchContextWorker.java

```diff
@@ -0,0 +1,150 @@
+        }
+
+        final OpenSearchIndexProgressState openSearchIndexProgressState = openSearchIndexProgressStateOptional.get();
+
+        final SearchConfiguration searchConfiguration = openSearchSourceConfiguration.getSearchConfiguration();
+        SearchWithSearchAfterResults searchWithSearchAfterResults = null;
+
+        // todo: Pass query and sort options from SearchConfiguration to the search request
+        do {
+            try {
+                searchWithSearchAfterResults = searchAccessor.searchWithoutSearchContext(NoSearchContextSearchRequest.builder()
+                                .withIndex(indexName)
+                                .withPaginationSize(searchConfiguration.getBatchSize())
+                                .withSearchAfter(getSearchAfter(openSearchIndexProgressState, searchWithSearchAfterResults))
+                                .build());
+
+                searchWithSearchAfterResults.getDocuments().stream().map(Record::new).forEach(record -> {
+                    try {
+                        bufferAccumulator.add(record);
+                    } catch (Exception e) {
+                        LOG.error("Failed writing OpenSearch documents to buffer. The last document created has document id '{}' from index '{}' : {}",
+                                record.getData().getMetadata().getAttribute(DOCUMENT_ID_METADATA_ATTRIBUTE_NAME),
+                                record.getData().getMetadata().getAttribute(INDEX_METADATA_ATTRIBUTE_NAME), e.getMessage());
+                    }
+                });
+            } catch (final Exception e) {
+                LOG.error("Received an exception while searching with PIT for index '{}'", indexName);
+                throw new RuntimeException(e);
+            }
+
+            openSearchIndexProgressState.setSearchAfter(searchWithSearchAfterResults.getNextSearchAfter());
+            sourceCoordinator.saveProgressStateForPartition(indexName, openSearchIndexProgressState);
+        } while (searchWithSearchAfterResults.getDocuments().size() == searchConfiguration.getBatchSize());
+
+        try {
+            bufferAccumulator.flush();
+        } catch (final Exception e) {
+            LOG.error("Failed writing remaining OpenSearch documents to buffer due to: {}", e.getMessage());
+        }
+    }
+
+    private OpenSearchIndexProgressState initializeProgressState() {
+        return new OpenSearchIndexProgressState();
+    }
+
+    private List<String> getSearchAfter(final OpenSearchIndexProgressState openSearchIndexProgressState, final SearchWithSearchAfterResults searchWithSearchAfterResults) {
+        if (Objects.isNull(searchWithSearchAfterResults) && Objects.isNull(openSearchIndexProgressState.getSearchAfter())) {
+            return null;
+        }
+
+        if (Objects.isNull(searchWithSearchAfterResults) && Objects.nonNull(openSearchIndexProgressState.getSearchAfter())) {
```

Comment:
You should combine these two if statements otherwise it is little confusing.

Reply:
I don't understand what is confusing about it.

Reply:
Oh I think that second if statement can be simplified to just `if (Objects.nonNull(openSearchIndexProgressState.getSearchAfter())`

Reply:
It is not clear from the code that line at 148  `searchWithSearchAfterResults` is not null. better way of writing would be 
```
if (Objects.isNull(searchWithSearchAfterResults)) {
      if (Objects.nonNull(openSearchIndexProgressState.getSearchAfter())( {
         return openSearchIndexProgressState.getSearchAfter();
     } else {
         return null;
     }
} 
```
This way we know after the code after the "if" block is the "else" block.